extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/S3UploadConfig.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.storage.s3.output;
+
+import org.apache.druid.guice.LazySingleton;
+import org.apache.druid.java.util.common.HumanReadableBytes;
+import org.apache.druid.java.util.common.logger.Logger;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * This class manages the configuration for uploading files to S3 in chunks.
+ * It tracks the current number of chunks written to local disk concurrently and ensures that the
+ * maximum number of chunks on disk does not exceed a specified limit at any given point in time.
+ */
+@LazySingleton
+public class S3UploadConfig
+{
+  /**
+   * The maximum chunk size based on injected values for {@link S3OutputConfig} and {@link S3ExportConfig} used for computing maximum number of chunks to save on disk at any given point in time.
+   * It is initialized to 5 MiB which is the minimum chunk size possible, denoted by {@link S3OutputConfig#S3_MULTIPART_UPLOAD_MIN_PART_SIZE_BYTES}.
+   */
+  private long chunkSize = new HumanReadableBytes("5MiB").getBytes();
+
+  /**
+   * An atomic counter to track the current number of chunks saved to local disk.
+   */
+  private final AtomicInteger currentNumChunks = new AtomicInteger(0);
```

Comment:
Why are we tracking something in a config? This doesn't seem very cool for a singleton. I would have expected it to be stateless. If this needs to be updated dynamically, it should mention that somewhere, `DynamicS3UploadConfig`, and a global instance of this should be avoided.
All the callers that need to share this must receive it explicitly instead of being injected via Guice. 
This seems error-prone if someone decides to use the storage connector down the line for some other task and they both share the same config.

Reply:
@LakshSingla The naming of the class could be better, please let me know if you have any alternate naming suggestions.

The intention here is to have a Singleton class which keeps a global state (across all instances of RetryableS3OutputStream) of how many chunks have been written to disk right now. It also updates the maxConcurrentNumChunks according to the chunk size.

The other option to achieve this would have been to have a static counter in RetryableS3OutputStream layer, but Karan pointed out that it's not a good practice, and hence we should go with a Singleton class.

Reply:
@LakshSingla I'm closing this comment in favor of this other comment which is kind of a superset: https://github.com/apache/druid/pull/16481#discussion_r1616615880

I'm working on the other comment, will leave that comment open until then for tracking. Hope that works!